source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc

```diff
@@ -0,0 +1,308 @@
+  }
+}
+
+void SPIFFEValidator::addClientValidationContext(SSL_CTX* ctx, bool) {
+  bssl::UniquePtr<STACK_OF(X509_NAME)> list(sk_X509_NAME_new(
+      [](const X509_NAME* const* a, const X509_NAME* const* b) -> int { return X509_NAME_cmp(*a, *b); }));
+
+  for (auto& ca : ca_certs_) {
+    X509_NAME* name = X509_get_subject_name(ca.get());
+
+    // Check for duplicates.
+    // Note that BoringSSL call only returns 0 or 1.
+    // OpenSSL can also return -1, for example on sk_find calls in an empty list 
+    if (sk_X509_NAME_find(list.get(), nullptr, name) >= 0) {
+      continue;
+    }
+
+    bssl::UniquePtr<X509_NAME> name_dup(X509_NAME_dup(name));
+    if (name_dup == nullptr || !sk_X509_NAME_push(list.get(), name_dup.release())) {
+      throw EnvoyException(absl::StrCat("Failed to load trusted client CA certificate"));
+    }
+  }
+  SSL_CTX_set_client_CA_list(ctx, list.release());
+}
+
+void SPIFFEValidator::updateDigestForSessionId(bssl::ScopedEVP_MD_CTX& md,
+                                               uint8_t hash_buffer[EVP_MAX_MD_SIZE],
+                                               unsigned hash_length) {
+  int rc;
+  for (auto& ca : ca_certs_) {
+    rc = X509_digest(ca.get(), EVP_sha256(), hash_buffer, &hash_length);
+    RELEASE_ASSERT(rc == 1, Utility::getLastCryptoError().value_or(""));
+    RELEASE_ASSERT(hash_length == SHA256_DIGEST_LENGTH,
+                   fmt::format("invalid SHA256 hash length {}", hash_length));
+    rc = EVP_DigestUpdate(md.get(), hash_buffer, hash_length);
+    RELEASE_ASSERT(rc == 1, Utility::getLastCryptoError().value_or(""));
+  }
+}
+
+int SPIFFEValidator::initializeSslContexts(std::vector<SSL_CTX*>, bool) {
+  return SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT;
+}
+
+int SPIFFEValidator::doVerifyCertChain(X509_STORE_CTX* store_ctx,
+                                       Ssl::SslExtendedSocketInfo* ssl_extended_info,
+                                       X509& leaf_cert, const Network::TransportSocketOptions*) {
+  if (!SPIFFEValidator::certificatePrecheck(&leaf_cert)) {
+    if (ssl_extended_info) {
+      ssl_extended_info->setCertificateValidationStatus(Envoy::Ssl::ClientValidationStatus::Failed);
+    }
+    stats_.fail_verify_error_.inc();
+    return 0;
+  }
+
+  auto trust_bundle = getTrustBundleStore(&leaf_cert);
+  if (!trust_bundle) {
+    if (ssl_extended_info) {
+      ssl_extended_info->setCertificateValidationStatus(Envoy::Ssl::ClientValidationStatus::Failed);
+    }
+    stats_.fail_verify_error_.inc();
+    return 0;
+  }
+
+  // Set the trust bundle's certificate store on the context, and do the verification.
+  bssl::UniquePtr<X509_STORE_CTX> verify_ctx(X509_STORE_CTX_new());
+  // We make a copy of X509_VERIFY_PARAMs in the store_ctx that we received as a parameter.
+  // This is a precaution mostly, as Envoy doesn't configure any X509_VERIFY_PARAMs.
+  // Note that there's no api to copy crls from one store_ctx to another; the assumption is that 
```

Comment:
This sentence is slightly hard to grok because of the double negation; if it is semantically correct probably not worth iterating on, but if this is accidental perhaps it is.

Reply:
Ugh, indeed, this is hard to understand, and I wrote it. Fixed.